### PR TITLE
`ultralytics 8.0.127` add FastSAM model

### PR DESCRIPTION
**Pull request by berry-ding**
### Issue
This Pull Request addresses the issue #3319.

### Changes Made
1. Added `FastSAM` module for prediction and validation.
2. Implemented mask selection based on box, point, and text prompt.
3. Added new code for visualizing masks.

 I have read the CLA Document and I sign the CLA

Please review this Pull Request at your convenience. Thank you!

### Summary
🖼️🎭🚀

This pull request adds a new feature to the ultralytics package: the FastSAM model for semantic segmentation. It introduces a new script `segpredict.py` that shows how to use the model and the FastSAMPrompt class to generate annotations. It also adds a new subpackage `fastsam` under `ultralytics/yolo` that contains the modules for the model, predictor, and prompt classes. It modifies the `__init__.py` file of the ultralytics package to import the FastSAM class. It also adds a new SegmentationValidator class that extends the DetectionValidator class to evaluate segmentation models.

> _`FastSAM` for segmen-_
> _tation: new script, package, modules_
> _Autumn leaves in masks_

### Walkthrough
*  Add a new script `segpredict.py` to demonstrate semantic segmentation with FastSAM ([link](https://github.com/ultralytics/ultralytics/pull/3390/files?diff=unified&w=0#diff-98b19c353cc1c2a1e2ff706f75e7c111d992d6863d687b729bd96505e0e91b30L1-R26))
*  Include FastSAM class in the list of available modules in `ultralytics/__init__.py` ([link](https://github.com/ultralytics/ultralytics/pull/3390/files?diff=unified&w=0#diff-b6202fa6738a34b56894dce00b24b8a4e4d7f4b1cbf1b04db6860f6b0f9635ceL10-R13))
*  Add a new subpackage `fastsam` under `ultralytics/yolo` ([link](https://github.com/ultralytics/ultralytics/pull/3390/files?diff=unified&w=0#diff-9af69d614cabda3505f9e594f59bdde5249abdddb227805150dbb9b14ea7377fR1-R9))
*  Define FastSAM class in `ultralytics/yolo/fastsam/model.py` ([link](https://github.com/ultralytics/ultralytics/pull/3390/files?diff=unified&w=0#diff-e903ab4ea12ad024cb11630009bc70836cd9429887e61b0301619092d2613e54R1-R107))
*  Define FastSAMPredictor class in `ultralytics/yolo/fastsam/predict.py` ([link](https://github.com/ultralytics/ultralytics/pull/3390/files?diff=unified&w=0#diff-fd7934273a4b88897f746c09649d169f9dc315332a5120ddb8d27ed3cfe90f0eL1-R40))
*  Define SegmentationValidator class in `ultralytics/yolo/fastsam/val.py` ([link](https://github.com/ultralytics/ultralytics/pull/3390/files?diff=unified&w=0#diff-0f2726e80af91eb1d966863eb54b2f82b28908ea373ad31ae21adb21a3946e63R1-R262))

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
Integration of FastSAM (Fast Segmentation-Aware Models) for single-class object detection and segmentation into the Ultralytics API.

### 📊 Key Changes
- Added easy-to-use Python API for FastSAM model integration into Python applications.
- Provided `predict` method implementation example for image object detection with FastSAM.
- Included `val` method example for model validation on a dataset.
- FastSAM only supports single-class detection and segmentation, requiring dataset category IDs conversion to 0.
- Updated Ultralytics package version to `8.0.127`.
- Expanded the set of importable objects with `FastSAM` in `__init__.py`.

### 🎯 Purpose & Impact
- Simplifies the process of implementing and using FastSAM within Python applications, making state-of-the-art segmentation more accessible.
- Users benefit from added features like retina masks, improved prediction processes, and validation procedures tailored for single-class scenarios.
- Impact includes enhanced performance, accessibility, and usability for those utilizing Ultralytics APIs for computer vision tasks.

### CODE REVIEW

**Comment by berry-ding** [comment]
我已阅读 CLA 文件并签署 CLA

**Comment by berry-ding** [comment]
recheck

**Comment by berry-ding** [comment]
recheck

**Comment by berry-ding** [comment]
I have read the CLA Document and I sign the CLA

**Comment by berry-ding** [comment]
recheck

**Comment by berry-ding** [comment]
I am new to Git and I am trying to integrate **FastSAM** into **YOLOv8**. Since **FastSAM** requires **CLIP** for _text prompt_ parsing, I attempted to add `git+https://github.com/openai/CLIP.git` to requirements.txt. However, I encountered `the pkg_resources.extern.packaging.requirements.InvalidRequirement: Parse error at "'+https:/'": Expected stringEnd `error during unit testing. Do you know how to resolve this issue?

**Comment by glenn-jocher** [comment]
@berry-ding that CLIP install should work. Here's an example of a similar install on L17:
https://github.com/ultralytics/ultralytics/blob/e93a5fbff648836a054f56a97c0afd523eb1795c/ultralytics/tracker/utils/matching.py#L9-L19

**Comment by berry-ding** [comment]
hi @glenn-jocher, thanks for your help! I have completed the code auto check. Regarding the issue of signing the CLA Document, since all three committers are myself, how should I solve this problem?

**Comment by glenn-jocher** [comment]
@berry-ding hello! Glad to hear that you've completed the code check. In regards to the CLA, if all three committers are yourself, then you can simply sign the CLA on behalf of all of them. Please leave a comment on the pull request stating that you have read the CLA document and that you sign the CLA on behalf of all committers, and that should be sufficient. Let me know if you have any further questions or concerns.

**Comment by berry-ding** [comment]
I have read the CLA document and that I sign the CLA on behalf of all committers

**Comment by berry-ding** [comment]
@glenn-jocher I don't have any more questions for now. Thank you!

**Comment by glenn-jocher** [comment]
@berry-ding you're welcome! It's great to hear that you don't have any more questions. If you do encounter any issues or have further questions in the future, don't hesitate to ask. We're always happy to help. Have a great day!

**Comment by glenn-jocher** [comment]
@berry-ding you're very welcome! If you have any further questions or concerns in the future regarding YOLOv8 or any other Ultralytics repository, don't hesitate to reach out. We're always happy to help!

**Comment by Zephyr69** [comment]
There is a visible border in the resulting image of everything mode. Is that expected? I check the original repo, and it seems it should not be the case?

**Comment by berry-ding** [comment]
Hello @Zephyr69 , the display of the edges is optional, and it is enabled by default. 
https://github.com/berry-ding/ultralytics/blob/7521f9cc6558e4c4ad5c5cd10570105ebaa69b11/ultralytics/yolo/fastsam/prompt.py#L91C1-L102C85
You can choose whether to enable the edges by setting the `withContours` parameter.
https://github.com/berry-ding/ultralytics/blob/7521f9cc6558e4c4ad5c5cd10570105ebaa69b11/segpredict.py#L30C1-L33C2

**Comment by Zephyr69** [comment]
> Hello @Zephyr69 , the display of the edges is optional, and it is enabled by default.
> 
> https://github.com/berry-ding/ultralytics/blob/7521f9cc6558e4c4ad5c5cd10570105ebaa69b11/ultralytics/yolo/fastsam/prompt.py#L91C1-L102C85
> 
> You can choose whether to enable the edges by setting the `withContours` parameter.
> 
> https://github.com/berry-ding/ultralytics/blob/7521f9cc6558e4c4ad5c5cd10570105ebaa69b11/segpredict.py#L30C1-L33C2

I mean the unmasked area near the edges of the image, not the contours of masks. Are those just false negatives?

**Comment by Zephyr69** [comment]
> @Zephyr69 Could you share the original images with me? I'd like to check whether it's an issue with the plotting code or a defect in the mask generation.

It's in this repo, 'ultralytics/assets/bus.jpg'.
This effect is more pronounced on the other image here:
zidane

**Comment by berry-ding** [comment]
@Zephyr69 , I've rerun the inference and visualized the results. You're right about the flaw in our model that you pointed out. I suspect this imprecision is due to poor localization. When dealing with large-scale bounding boxes, YOLOv8 might not be able to precisely locate arbitrary objects. Our method based on YOLOv8-seg involves detection first and then segmentation based on the detection results, so the quality of the bounding box generation significantly affects the segmentation outcome. Here are the results of my visualization of the bounding boxes.

test11

**Comment by Zephyr69** [comment]
> @Zephyr69 , I've rerun the inference and visualized the results. You're right about the flaw in our model that you pointed out. I suspect this imprecision is due to poor localization. When dealing with large-scale bounding boxes, YOLOv8 might not be able to precisely locate arbitrary objects. Our method based on YOLOv8-seg involves detection first and then segmentation based on the detection results, so the quality of the bounding box generation significantly affects the segmentation outcome. Here are the results of my visualization of the bounding boxes.
> 
> test11

I see. Is it possible to fix this then? Does it make sense to just append a bbox that encompasses the whole image?

**Comment by berry-ding** [comment]
@Zephyr69 ,Here's an engineering trick you can use: you can add a few pixels of white border to the original image, and then feed the resulting picture into the network. This way, you can more accurately locate objects at the edge of the image, as shown in the figure below.
test11

`from PIL import Image, ImageOps

def add_border(input_image_path, output_image_path, border_size=10):
    img = Image.open(input_image_path)
    border_img = ImageOps.expand(img, border=border_size, fill='white')
    border_img.save(output_image_path)

# usage
add_border('zidane.jpg', 'zidane2.jpg', 10)`

**Comment by glenn-jocher** [comment]
Hey @berry-ding, 
Thanks for bringing this issue to our attention! In YOLOv8, it is possible to encounter imprecise localization, especially when dealing with large-scale bounding boxes. This can result in unmasked areas near the edges of the image. One possible solution that you can try is to add a few pixels of a white border to the original image before feeding it into the network. By doing this, you can improve the accuracy of object localization, as shown in the example image you provided. 

You can achieve this by using the Pillow library in Python. Here's a simple code snippet that demonstrates how to add a white border to an image:

```python
from PIL import Image, ImageOps

def add_border(input_image_path, output_image_path, border_size=10):
    img = Image.open(input_image_path)
    border_img = ImageOps.expand(img, border=border_size, fill='white')
    border_img.save(output_image_path)

# Usage:
add_border('zidane.jpg', 'zidane2.jpg', 10)
```

This code will add a 10-pixel white border to the image "zidane.jpg" and save the modified image as "zidane2.jpg".

By using this method, you can more accurately locate objects at the edges of the image and potentially reduce the unmasked areas in the segmentation results.

Please let us know if you have any further questions or need more assistance. We're here to help!

**Comment by berry-ding** [comment]
Hi @glenn-jocher , I am wondering how I could improve this PR, or if it's fine as it is and I can move on to other tasks. I'm not quite familiar with the rules of submitting PRs, so could you please elaborate?

**Comment by Zephyr69** [comment]
@berry-ding Can we add some bandaid-fix to the post-processing so the model can deal with as much unmasking issue as possible?

First add two util functions:
def adjust_bboxes_to_image_border(boxes, image_shape, threshold=20):
    '''Adjust bounding boxes to stick to image border if they are within a certain threshold.

    Args:
    boxes: (n, 4)
    image_shape: (height, width)
    threshold: pixel threshold

    Returns:
    adjusted_boxes: adjusted bounding boxes
    '''

    # Image dimensions
    H, W = image_shape

    # Adjust boxes
    boxes[:, 0] = torch.where(boxes[:, 0] < threshold, 0, boxes[:, 0])  # x1
    boxes[:, 1] = torch.where(boxes[:, 1] < threshold, 0, boxes[:, 1])  # y1
    boxes[:, 2] = torch.where(boxes[:, 2] > W - threshold, W, boxes[:, 2])  # x2
    boxes[:, 3] = torch.where(boxes[:, 3] > H - threshold, H, boxes[:, 3])  # y2

    return boxes


def bbox_iou(box1, boxes, iou_thres=0.9, image_shape=(640, 640), raw_output=False):
    '''Compute the Intersection-Over-Union of a bounding box with respect to an array of other bounding boxes.

    Args:
    box1: (4, )
    boxes: (n, 4)

    Returns:
    high_iou_indices: Indices of boxes with IoU > thres
    '''
    boxes = adjust_bboxes_to_image_border(boxes, image_shape)
    # obtain coordinates for intersections
    x1 = torch.max(box1[0], boxes[:, 0])
    y1 = torch.max(box1[1], boxes[:, 1])
    x2 = torch.min(box1[2], boxes[:, 2])
    y2 = torch.min(box1[3], boxes[:, 3])

    # compute the area of intersection
    intersection = (x2 - x1).clamp(0) * (y2 - y1).clamp(0)

    # compute the area of both individual boxes
    box1_area = (box1[2] - box1[0]) * (box1[3] - box1[1])
    box2_area = (boxes[:, 2] - boxes[:, 0]) * (boxes[:, 3] - boxes[:, 1])

    # compute the area of union
    union = box1_area + box2_area - intersection

    # compute the IoU
    iou = intersection / union  # Should be shape (n, )
    if raw_output:
        if iou.numel() == 0:
            return 0
        return iou

    # get indices of boxes with IoU > thres
    high_iou_indices = torch.nonzero(iou > iou_thres).flatten()

    return high_iou_indices

Then, after we got nms results p:
full_box = torch.zeros_like(p[0][0])
full_box[2], full_box[3], full_box[4], full_box[6:] = img.shape[3], img.shape[2], 1.0, 1.0
full_box = full_box.view(1, -1)
critical_iou_index = bbox_iou(full_box[0][:4], p[0][:, :4], iou_thres=0.9, image_shape=img.shape[2:])
p[0][critical_iou_index] = full_box

The results are like this:
bus
zidane

**Comment by glenn-jocher** [comment]
Hi @berry-ding, thank you for your contribution! Your proposal to address the unmasking issue is definitely interesting.

To provide a bandaid-fix for the post-processing step, you can add two utility functions. The first function, `adjust_bboxes_to_image_border`, adjusts the bounding boxes so that they stick to the image borders if they are within a certain threshold. This ensures that objects near the edges of the image are better captured. The second function, `bbox_iou`, computes the Intersection over Union (IoU) between a given bounding box and other bounding boxes. By using these functions, you can identify boxes with high IoU values and adjust them accordingly.

Here's how you can implement the proposed changes:

1. Add the following utility functions:

```python
import torch

def adjust_bboxes_to_image_border(boxes, image_shape, threshold=20):
    h, w = image_shape
    boxes[:, 0] = torch.where(boxes[:, 0] < threshold, 0, boxes[:, 0])  # x1
    boxes[:, 1] = torch.where(boxes[:, 1] < threshold, 0, boxes[:, 1])  # y1
    boxes[:, 2] = torch.where(boxes[:, 2] > w - threshold, w, boxes[:, 2])  # x2
    boxes[:, 3] = torch.where(boxes[:, 3] > h - threshold, h, boxes[:, 3])  # y2
    return boxes

def bbox_iou(box1, boxes, iou_thres=0.9, image_shape=(640, 640), raw_output=False):
    boxes = adjust_bboxes_to_image_border(boxes, image_shape)
    x1 = torch.max(box1[0], boxes[:, 0])
    y1 = torch.max(box1[1], boxes[:, 1])
    x2 = torch.min(box1[2], boxes[:, 2])
    y2 = torch.min(box1[3], boxes[:, 3])
    intersection = (x2 - x1).clamp(0) * (y2 - y1).clamp(0)
    box1_area = (box1[2] - box1[0]) * (box1[3] - box1[1])
    box2_area = (boxes[:, 2] - boxes[:, 0]) * (

**Comment by berry-ding** [comment]
Hi @Zephyr69 @glenn-jocher , great idea!!!!!!!!!!! I've added the code to my PR. Note that after selecting the largest candidate box, we only need to modify the size of the bounding box, and the existing confidence and prototype parameters of the candidate box should remain unchanged. I found that this does not degrade the performance. Also, problems will arise in the point prompt mode if it's directly set to 1.0.
```
        full_box = torch.zeros_like(p[0][0])
        full_box[2], full_box[3], full_box[4], full_box[6:] = img.shape[3], img.shape[2], 1.0, 1.0
        full_box = full_box.view(1, -1)
        critical_iou_index = bbox_iou(full_box[0][:4], p[0][:, :4], iou_thres=0.9, image_shape=img.shape[2:])
        if critical_iou_index.numel() != 0:
            full_box[0][4] = p[0][critical_iou_index][:,4]
            full_box[0][6:] = p[0][critical_iou_index][:,6:]
            p[0][critical_iou_index] = full_box
```

**Comment by glenn-jocher** [comment]
@berry-ding this looks great now! Can you just resolve the precommit issue with importing all, i.e. in https://results.pre-commit.ci/run/github/535360445/1688024654.tuJnNHx4Sq-ubdm-UUofoA

**Comment by berry-ding** [comment]
Hi @glenn-jocher , I have fixed the issue.

**Comment by AyushExel** [comment]
@berry-ding I think you'll need to sign the CLA too

**Comment by berry-ding** [comment]
I have read the CLA Document and I sign the CLA

**Comment by berry-ding** [comment]
Hi @AyushExel , I have already signed the CLA agreement. However, my previous commits did not link to my email, resulting in my commits being split into three committers. I am able to represent all committers in signing the CLA agreement as I only have one account. Could you please advise on how I can resolve this issue?

**Comment by AyushExel** [comment]
@glenn-jocher we can probably merge this without CLA passing as all relevant people have signed

**Comment by glenn-jocher** [comment]
Hi @berry-ding. Thank you for updating the docs and adding usage examples. Your contribution is greatly appreciated! If you have any more questions or need further assistance, feel free to reach out.

**Comment by glenn-jocher** [comment]
@berry-ding ok I've made a few small changes. Looks good. Waiting for CI and then should be good to merge.

**Comment by glenn-jocher** [comment]
recheck

**Comment by glenn-jocher** [comment]
@berry-ding PR merged! Thank you for your FastSAM contribution! :)
